src/Parallel/Algorithm.hpp

```diff
@@ -55,6 +57,152 @@ template <typename ParallelComponent, typename PhaseDepActionList>
+///  entering any global sync phases.
+enum AlgorithmExecution { Continue, Pause, SleepForSyncPhases, Halt };
+
+/// A collection of information about the control-flow for the algorithm and
+/// phase selection
+///
+/// \details This struct is used in one of the options for the iterable action
+/// return, for which any combination of the member variables may be set to
+/// transmit control flow information back to the algorithm. The member
+/// variables are:
+/// - `execution_flag` : an `Parallel::AlgorithmExecution` indicating whether
+/// the execution of the algorithm should proceed. Default
+/// `Parallel::AlgorithmExecution::Continue` continues execution of the
+/// algorithm normally.
+/// - `next_action_index` : an optional `size_t` indicating the next action
+/// index in the current phase to jump to. Default `std::nullopt` continues
+/// execution at the next iterable action normally.
+/// - `global_sync_phases` : an optional `std::unordered_set<typename
+/// Metavariables::Phase>` of global sync phases to request. Default
+/// `std::nullopt` requests no global sync phases.
+template <typename Metavariables>
+struct AlgorithmControl {
+  AlgorithmExecution execution_flag = AlgorithmExecution::Continue;
+  std::optional<size_t> next_action_index;
+  std::optional<std::unordered_set<typename Metavariables::Phase>>
+      global_sync_phases;
+};
+
+namespace detail {
+
+// obtains the Chare type for the component associated with a template
+// instantiation of the `AlgorithmImpl`, for choosing the appropriate type from
+// which to inherit.
+template <typename Component, typename ChareType>
+struct get_charm_base_class;
+
+template <typename Component>
+struct get_charm_base_class<Component, Parallel::Algorithms::Array> {
+  using type = CBase_AlgorithmArray<
+      Component, typename get_array_index<
+                     Parallel::Algorithms::Array>::template f<Component>>;
+};
+
+template <typename Component>
+struct get_charm_base_class<Component, Parallel::Algorithms::Group> {
+  using type = CBase_AlgorithmGroup<
+      Component, typename get_array_index<
+                     Parallel::Algorithms::Group>::template f<Component>>;
+};
+
+template <typename Component>
+struct get_charm_base_class<Component, Parallel::Algorithms::Nodegroup> {
+  using type = CBase_AlgorithmNodegroup<
+      Component, typename get_array_index<
+                     Parallel::Algorithms::Nodegroup>::template f<Component>>;
+};
+
+template <typename Component>
+struct get_charm_base_class<Component, Parallel::Algorithms::Singleton> {
+  using type = CBase_AlgorithmSingleton<
+      Component, typename get_array_index<
+                     Parallel::Algorithms::Singleton>::template f<Component>>;
+};
+
+template <typename Component>
+using get_charm_base_class_t =
+    typename get_charm_base_class<Component,
+                                  typename Component::chare_type>::type;
```

Comment:
I don't see `get_charm_base_class` being used elsewhere in the PR. Is it needed?

Furthermore, isn't this information now encapsulated by `Component::cbase` (if I've correctly understood that Component is one of the structs in Parallel::Algorithms?)?

Reply:
ah, shoot -- yeah, you're right, this is a change that my `serialize_algorithm` alterations means this entire type trait should just be removed.